cmd/james/serve.go

```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"regexp"
+	"syscall"
+
+	"github.com/bwmarrin/discordgo"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/spf13/cobra"
+
+	"github.com/itfactory-tm/thomas-bot/pkg/command"
+	"github.com/itfactory-tm/thomas-bot/pkg/commands/hello"
+	"github.com/itfactory-tm/thomas-bot/pkg/discordha"
+)
+
+func init() {
+	rootCmd.AddCommand(NewServeCmd())
+}
+
+type serveCmdOptions struct {
+	Token    string
+	Prefix   string `default:"bob"`
+
+	EtcdEndpoints      []string `envconfig:"ETCD_ENDPOINTS"`
+
+	commandRegex *regexp.Regexp
+	dg           *discordgo.Session
+	ha           *discordha.HA
+	handlers     []command.Interface
+
+	onMessageCreateHandlers     map[string][]func(*discordgo.Session, *discordgo.MessageCreate)
+	onMessageEditHandlers       map[string][]func(*discordgo.Session, *discordgo.MessageUpdate)
+	onMessageReactionAddHandler []func(*discordgo.Session, *discordgo.MessageReactionAdd)
+	onGuildMemberAddHandler     []func(*discordgo.Session, *discordgo.GuildMemberAdd)
+}
+
+// NewServeCmd generates the `serve` command
+func NewServeCmd() *cobra.Command {
+	s := serveCmdOptions{}
+	c := &cobra.Command{
+		Use:     "serve",
+		Short:   "Run the server",
+		Long:    `This connects to Discord and handle all game related events`,
+		RunE:    s.RunE,
+		PreRunE: s.Validate,
+	}
+
+	// TODO: switch to viper
+	err := envconfig.Process("thomasbob", &s)
+	if err != nil {
+		log.Fatalf("Error processing envvars: %q\n", err)
+	}
+
+	return c
+}
+
+func (s *serveCmdOptions) Validate(cmd *cobra.Command, args []string) error {
+	if s.Token == "" {
+		return errors.New("No token specified")
+	}
+
+	s.RegisterHandlers()
+
+	return nil
+}
+
+func (s *serveCmdOptions) RunE(cmd *cobra.Command, args []string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	s.commandRegex = regexp.MustCompile(s.Prefix + `!(\w*)\b`)
+
+	dg, err := discordgo.New("Bot " + s.Token)
+	if err != nil {
+		return fmt.Errorf("error creating Discord session: %w", err)
+	}
+
+	//Creates intent error for some reason
+	//dg.Identify.Intents = discordgo.MakeIntent(discordgo.IntentsAll)
+
+	s.ha, err = discordha.New(discordha.Config{
+		Session:       dg,
+		HA:            len(s.EtcdEndpoints) > 0,
+		EtcdEndpoints: s.EtcdEndpoints,
+		Context:       ctx,
+	})
+	if err != nil {
+		return fmt.Errorf("error creating Discord HA: %w", err)
+	}
+
+	// TODO: Register handlers
+	dg.AddHandler(s.onMessage)
+
+	err = dg.Open()
+	if err != nil {
+		return fmt.Errorf("error opening connection: %w", err)
+	}
+	defer dg.Close()
+
+	dg.UpdateStreamingStatus(0, fmt.Sprintf("Awaiting your command, use me with bob!"), "")
```

Comment:
can we re-add the revision here? was useful for me to see when updates were complete